(PUP-8971) add encrypted data type (for puppet 6.1)

hlindberg opened the pull request:
This adds the Encrypted data type.

puppetcla commented:
CLA signed by all contributors.

jsane commented on lib/puppet/pops/types/p_encrypted_type.rb:
```diff
@@ -0,0 +1,354 @@
+
+      # Check if this was for "this recipient"
+      #
+      unless fingerprint == host.certificate.fingerprint
```
One of other purposes this serves aside from the intended recipient check, though not completely, is to check the integrity of the encrypted portions. We may want to consider having an explicit integrity check by 'signing' the hashes of these encrypted portions using the server key.  Or perhaps the complete catalog if it is easier and with less space requirements.
Also the intended recipient check can moved earlier so we don't spend time decrypting the crypt part (unless the format is forcing you to). You could also include the intended recipient in plain text - it is meant to be a quick sanity check since actual decryption attempt using wrong key fail anyway (there are ways to detect if the decryption failed if RSA decryption does not report errors). 
Agree with Josh - since we are using a different IV for every secret no need to create a new key. Vice versa - if you are creating a new key every time then IV should not be required.

joshcooper commented:
This is https://github.com/puppetlabs/puppet/pull/6899 just targeting master.

joshcooper commented:
Copying my comments from the earlier PR:

1. It is desirable to generate a session key per catalog compile, not per secret. I don't think it's blocking.
1. Since we're targeting `master` which requires ruby >= 2.3, we can drop the `if RUBY_VERSION < "2.1.0"` branch.
1. We should implement integrity protection to protect against bit-flipping. The use of a fingerprint doesn't cover all of the data. One option is to use ciphers that support AEAD. It generates a tag as part of the encryption process, and can be used to verify the ciphertext is unmodified.
1. The decryptor should not make assumptions about the algorithm used to encrypt (since those can happen on different systems at different times). So we can't assume the `IV` is 16 bytes (eg `buffer[0..15], buffer[16..-1]]`). The encryptor needs to provide all of the public inputs that the decryptor will need later. This includes `IV`, `algorithm`, and if using AEAD, `tag`.
1. OpenSSL provides `EVP_Seal*` methods to derive a symmetric session key, encrypt the secret using the session key, and encrypt the session key with the recipients' public key. I don't think it's accessible via Ruby, but would be good to confirm.

hlindberg commented:
Lots of good input - I will need help though to make those changes, and it seems we need to discuss what exactly we want to do. 
Will start by rebasing this PR onto master.

hlindberg commented:
1. Keeping a session key in the catalog - can be done but is a bit tricky also considering use cases where there is no catalog. @joshcooper said "not block", which I take as "does not have to be done in first release of Encrypted data type"
2. Fixed, the backwards compatibility for `Error` `cause` for Ruby < 2.1.0 is now dropped
3. Protect against bit-flipping; isn't this important for the entire catalog? (If I can bit flip encrypted data then I can also change any data in the catalog). If supporting `AEAD`, should the user ask for it, or should it always be used if algorithm supports it? Alternatively possible to opt out if on by default when supported?
4. The `Encrypted` data type has a `format` which is the combined cipher-name and serialization format name, adding additional attributes is possible, but I would then break them out into individual optional attributes, for example `format`, `cipher`, `iv_length`, and `tag`.
5. Will investigate

hlindberg commented:
On topic (5) - This [stackoverflow-post](https://stackoverflow.com/questions/5212213/ruby-equivalent-of-php-openssl-seal) answers that SEAL is not in the Ruby OpenSSL implementation, but it points to [this gist](https://gist.github.com/marcbowes/1899731) that shows how to do a SEAL using RSA (asymmetric) - which is pretty much how it is already implemented only that the current implementation is using the same cipher algorithm also for the session key.

Not sure how to proceed.

hlindberg commented:
Need to discuss how to best proceed - the use of `AEAD` (essentially `GCM`) requires an IV which can be static (`0`) if key is guaranteed to be different each time, and if it is not, a context + counter or a random IV (not recommended) must be used. Is the use of a random key (per encryption) guaranteed to be (sufficiently) unique?

Anything that requires a context in which encrypt/decrypt takes place makes the process more complicated. In a non catalog scenario, how do you get the key across - what data type is that?

When using GCM, what is a suitable tag to use? The cert fingerprint? And send that as clear text? (Is it ok to reveal that? It identifies the target, and could that be considered sensitive?)

Advice from @jsane above says if key is different each time, then no IV is needed, but with GCM there is always an IV (and preferably (?) it should use unique key).

Then mix the above with use of `SEAL` - as per the gist above? To encrypt the random key with RSA?

jtappa commented:
Just checking in, there's been no updates or changes to this PR for a year, is this still a data type we'd like to implement? If so what work is left that we need to do to implement it/how can we support you @hlindberg? Or, should we close the PR for now, until we have a path forward?

hlindberg commented:
I think this is of value. I got suggestions/feedback regarding how encryption should work that I did not know how to handle right away. I then got busy with other things.
Basically, someone needs to pick up answer the questions I raised in my last comment above and take it from there. Happy to help that person out and talk about all the things I implemented but I don't have the time now to learn the stuff that was pointed out as being required for this to be a good solution. OTOH - what is implemented is better than not having this at all IMO. It can always be improved...

trevor-vaughan commented:
I also think that this is of value. Having this be an intrinsic part of the language would be preferable to an add-on solution.

jtappa commented:
Rekicking Travis

jtappa commented:
Closing this for now, as we don't have the bandwidth to move forward with it and it has grown stale.

We can always reopen if we get the time.
